File: typescript/y2025/d10-factory.ts
type Machine = {
    target: string;
    buttons: number[][];
    joltage: number[];
}

export function parse(input: string): Machine[] {
    return input.split('\n').map(line => {
        // Extract target from [...]
        const targetMatch = line.match(/\[([^\]]+)\]/);
        const target = targetMatch ? targetMatch[1] : '';

        // Extract buttons from (...) patterns
        const buttonMatches = [...line.matchAll(/\(([^)]+)\)/g)];
        const buttons = buttonMatches.map(match =>
            match[1].split(',').map(n => Number.parseInt(n.trim()))
        );

        // Extract joltage from {...}
        const joltageMatch = line.match(/\{([^}]+)\}/);
        const joltage = joltageMatch
            ? joltageMatch[1].split(',').map(n => Number.parseInt(n.trim()))
            : [];

        return { target, buttons, joltage };
    });
}

type MachineState = {
    indicator: string;
    joltage: number[];
    pressedButtons: Set<number>;
    steps: number;
};

const initialize = (length: number): MachineState => ({
    indicator: '.'.repeat(length),
    joltage: Array.from({ length }).fill(0) as number[],
    pressedButtons: new Set(),
    steps: 0
});

// generate key for deduplication
const stateKey = (pattern: string, pressed: Set<number>): string =>
    `${pattern}|${[...pressed].sort().join(',')}`;

const togglePositions = (pattern: string, positions: number[]): string =>
    pattern.split('')
        .map((char, idx) => positions.includes(idx) ? (char === '#' ? '.' : '#') : char)
        .join('');

export const incrementJoltage = (joltage: number[], positions: number[]): number[] =>
    joltage.map((value, idx) => positions.includes(idx) ? value + 1 : value);

const getUnpressedButtons = (totalButtons: number, pressed: Set<number>): number[] =>
    Array.from({ length: totalButtons }, (_, i) => i)
        .filter(i => !pressed.has(i));

const pressButton = (current: MachineState, buttonIndex: number, button: number[]): MachineState => ({
    indicator: togglePositions(current.indicator, button),
    joltage: incrementJoltage(current.joltage,button),
    pressedButtons: new Set([...current.pressedButtons, buttonIndex]),
    steps: current.steps + 1
});

const getNextIndicatorState = (
    current: MachineState,
    buttons: number[][],
    visited: Set<string>
): MachineState[] => {
    return getUnpressedButtons(buttons.length, current.pressedButtons)
        .map(buttonIndex => pressButton(current, buttonIndex, buttons[buttonIndex]))
        .filter(state => {
            const key = stateKey(state.indicator, state.pressedButtons);
            if (visited.has(key)) return false;
            visited.add(key);
            return true;
        });
};

/**
 * Find the minimum number of button presses needed to reach the target using BFS.
 * @param machine - The machine with target and available buttons
 * @returns The minimum number of button presses, or -1 if impossible
 */
export function minButtonPresses(machine: Machine): number {
    const initial = initialize(machine.target.length);
    if (initial.indicator === machine.target) return 0;

    const visited = new Set([stateKey(initial.indicator, initial.pressedButtons)]);
    const queue: MachineState[] = [initial];

    while (queue.length > 0) {
        const current = queue.shift();
        if (!current) break;

        const nextStates = getNextIndicatorState(current, machine.buttons, visited);

        // Check if any next state reaches the target
        const solution = nextStates.find(state => state.indicator === machine.target);
        if (solution) return solution.steps;

        queue.push(...nextStates);
    }

    return -1; // Target unreachable
}


export function partOne(input: string) {
    const machines = parse(input);
    return machines.reduce((acc, machine) => acc + minButtonPresses(machine), 0);
}

export function partTwo(input: string) {
    const machines = parse(input);
    for(const machine of machines) {
        console.log(machine);
    }
    return 0;
}

File: go/y2025/d10_factory.go
package main

import (
	_ "embed"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

//go:embed inputs/d10.txt
var puzzleInput string

/*

The manual describes one machine per line.
Each line contains
- a single indicator light diagram in [square brackets],
- one or more button wiring schematics in (parentheses),
- and joltage requirements in {curly braces}.

   [.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
   [...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}
   [.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}

indicator lights are all initially off
*/

type Button []int

func (b Button) mask() uint {
	var mask uint
	for _, n := range b {
		mask |= 1 << n
	}
	return mask
}

type Indicator string

func (i Indicator) mask() uint {
	var mask uint
	for idx, r := range i {
		if r == '#' {
			mask |= 1 << idx
		}
	}
	return mask
}

type State struct {
	indicator uint
	pressed   uint
	steps     int
}

type Machine struct {
	target  string
	buttons []Button
	joltage []int
}

func newMachine(target string, buttons []Button, joltage []int) Machine {
	machine := Machine{target, buttons, joltage}
	return machine
}

var targetPattern = regexp.MustCompile(`\[([^\]]+)\]`)
var buttonPattern = regexp.MustCompile(`\(([^)]+)\)`)
var joltagePattern = regexp.MustCompile(`\{([^}]+)\}`)

func parseNumbers(s string) []int {
	parts := strings.Split(s, ",")
	numbers := make([]int, len(parts))
	for i, p := range parts {
		n, err := strconv.Atoi(strings.TrimSpace(p))
		if err != nil {
			panic(err)
		}
		numbers[i] = n
	}
	return numbers
}

func parseMachine(line string) Machine {
	target := ""
	if match := targetPattern.FindStringSubmatch(line); match != nil {
		target = match[1]
	}

	buttonMatches := buttonPattern.FindAllStringSubmatch(line, -1)
	buttons := make([]Button, len(buttonMatches))
	for i, match := range buttonMatches {
		buttons[i] = parseNumbers(match[1])
	}

	joltage := []int{}
	if match := joltagePattern.FindStringSubmatch(line); match != nil {
		joltage = parseNumbers(match[1])
	}

	return newMachine(target, buttons, joltage)
}

func parse(input string) []Machine {
	lines := strings.Split(strings.TrimSpace(input), "\n")
	machines := make([]Machine, len(lines))
	for i, line := range lines {
		machines[i] = parseMachine(line)
	}
	return machines
}

func minButtonPresses(m Machine) int {
	target := Indicator(m.target).mask()
	var initial State

	if initial.indicator == target {
		return 0
	}
	visited := map[State]bool{}
	visited[initial] = true

	queue := []State{initial}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		for idx, button := range m.buttons {
			buttonBit := uint(1) << idx
			if current.pressed&buttonBit != 0 {
				continue
			}

			next := State{
				indicator: current.indicator ^ button.mask(),
				pressed:   current.pressed | buttonBit,
				steps:     current.steps + 1,
			}

			if next.indicator == target {
				return next.steps
			}

			if !visited[next] {
				visited[next] = true
				queue = append(queue, next)
			}
		}
	}

	panic("target unreachable")
}

func partOne(input string) int {
	total := 0
	for _, m := range parse(input) {
		total += minButtonPresses(m)
	}
	return total
}

func main() {
	fmt.Println(partOne(puzzleInput))
}
